File: Longest_Substring_Without_Repeating_Characters/sol3.cpp
/* -.-.-.-.-.-.-.-.-.-.-.-.-.-.-.-.-.-.-.-.
* File Name : sol3.cpp
* Purpose : refined code
* Creation Date : 25-01-2014
* Last Modified : Sat Jan 25 18:26:51 2014
* Created By : wdd 
_._._._._._._._._._._._._._._._._._._._._.*/
#include "general.h"
    int lengthOfLongestSubstring(string s) {
        map<char, int> mp;
        int res = 0, start = 0, len;
        for(int i = 0; i < s.length() + 1; ++ i){
            len = i - start;
            if(i == s.length()) break;
            map<char, int>::iterator itr = mp.find(s[i]);
            if(itr != mp.end()){
                if(itr->second >= start){
                    if(len > res)res = len;
                    start = itr->second + 1;
                }
                itr->second = i;
            }else
                mp[s[i]] = i;
        }
        if(len > res) res = len;
    	return res;
    }
int main(int argc, const char* argv[])
{
	return 0;
}

File: Longest_Substring_Without_Repeating_Characters/sol3.py
#! /usr/bin/python
# --------------------------------------- 
# File Name : sol3.py
# Creation Date : 24-11-2014
# Last Modified : Mon Nov 24 11:01:15 2014
# Created By : wdd 
# --------------------------------------- 
import sys, re, string, pickle
from itertools import *
class Solution:
    # @return an integer
    def lengthOfLongestSubstring(self, s):
        dict_book, max_len, start = {}, 0, 0
        for i, val in enumerate(s):
            if val in dict_book and dict_book[val] >= start:
                tmp_len = i - start
                if tmp_len > max_len: max_len = tmp_len
                start = dict_book[val] + 1
            dict_book[val] = i
        tmp_len = len(s) - start
        if tmp_len > max_len:
            max_len = tmp_len
        return max_len
if __name__ == "__main__": main()
